doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+# Catalog-only mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets users boot into Materialize and take actions to remediate the issue,
+e.g. drop the OOMing view. To accomplish this, we'll introduce a new boot mode
+called "catalog-only mode" that starts Materialize as normal with the exception
+of not shipping any dataflows for user-created objects, such as materialized
+views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+## Goals
+
+- Support booting in catalog-only mode, which will not:
+    - Ship the dataflow for any user-created object
+    - Support querying user-created objects, such as views
+    - Support inserting into tables
+
+    Any other action should be allowed, such as creating, modifying, or dropping
```

Comment:
> What will MZ do when the user creates a new materialized view in catalog-only mode? Will that be surprising or confusing to the user?

The new view will be in the same state as the other views when they boot: their indexes will not exist.

There's a chance that this will be surprising to users, and is something we'll need to address in the docs, whether they're internal or external. We might also consider setting some kind of additional flag on the create statements' responses in catalog-only mode that expressly names that any indexes the user might expect to get created were not and will need to be manually enabled.

A note that I think the alternative behavior of creating the index might also surprise users who expect this to behave like a safe mode.

> Also, how do you envision a user would use catalog-only mode? What specific steps would they take to get out of the OOM loop that motivated their restarting MZ in catalog-only mode? Would it make things less error-prone to disallow operations that are not needed to restore MZ to a working state?

This is a great question: I heard this as an ask from the cloud team to aid them in helping users resume a cluster's operation after creating a troubled view. The expected use case was logging into a customer's box in catalog only mode, looking at the system tables to determine which dataflow caused the OOM, and then dropping the offending view.

To make catalog-only mode useful for end-users, we'll need a lot more documentation and education around using the system's built-in views to make this kind of assessment. My initial sense was that this was out of scope and could be built up iteratively if we did want this to become a first-class debugging tool for users.

Prohibiting other action is an interesting idea and something that hadn't occurred to me during the design, given that I was building for an internal team. We could certainly add it somewhat easily--entering the catalog into a type of read-only mode that only let you create e.g. temporary diagnostic views.